src/operation/polygonize/BuildArea.cpp

```diff
@@ -0,0 +1,246 @@
+#include <geos/operation/union/CascadedPolygonUnion.h>
+#include <geos/geom/Geometry.h>
+#include <geos/geom/GeometryCollection.h>
+#include <geos/geom/GeometryFactory.h>
+#include <geos/geom/LineString.h>
+#include <geos/geom/Polygon.h>
+#include <geos/geom/MultiPolygon.h>
+
+//#define DUMP_GEOM
+#ifdef DUMP_GEOM
+#include <geos/io/WKTWriter.h>
+#endif
+
+// std
+#include <algorithm>
+#include <vector>
+
+#ifdef _MSC_VER
+#pragma warning(disable:4355)
+#endif
+
+using namespace std;
+using namespace geos::geom;
+
+namespace geos {
+namespace operation { // geos.operation
+namespace polygonize { // geos.operation.polygonize
+
+struct Face {
+    const geom::Polygon* poly = nullptr;
+    std::unique_ptr<geom::Geometry> env; // envelope
+    double envarea = 0.0; // envelope area
+    Face* parent = nullptr; /* if this face is an hole of another one, or NULL */
+
+    size_t countParents() const {
+        const Face* f = this;
+        size_t pcount = 0;
+        while ( f->parent ) {
+            ++pcount;
+            f = f->parent;
+        }
+        return pcount;
+    }
+};
+
+static std::unique_ptr<Face> newFace(const geom::Polygon* p) {
+    auto f = std::unique_ptr<Face>(new Face());
+    f->poly = p;
+    f->env.reset(p->getEnvelope());
+    f->envarea = f->env->getArea();
+    return f;
+}
+
+struct CompareByEnvarea {
+
+    bool operator()(const std::unique_ptr<Face> &a,
+                    const std::unique_ptr<Face> &b) const {
+        return a->envarea > b->envarea;
+    }
+};
+
+/* Find holes of each face */
+static void findFaceHoles(std::vector<std::unique_ptr<Face>>& faces) {
+
+    /* We sort by decreasing envelope area so that we know holes are only
+     * after their shells */
+    std::sort(faces.begin(), faces.end(), CompareByEnvarea());
+
+    const size_t nfaces = faces.size();
+    for( size_t i = 0; i < nfaces; ++i ) {
+        auto& f = faces[i];
+        const size_t nholes = f->poly->getNumInteriorRing();
+        for( size_t h = 0; h < nholes; h++ ) {
+            const auto hole = f->poly->getInteriorRingN(h);
+            for( auto j=i+1; j < nfaces; ++j ) {
+                auto& f2 = faces[j];
+                if( f2->parent ) {
+                    continue; /* hole already assigned */
+                }
+                const auto f2er = f2->poly->getExteriorRing();
+                /* TODO: can be optimized as the ring would have the
+                *       same vertices, possibly in different order.
+                *       maybe comparing number of points could already be
+                *       useful.
+                */
+                if( f2er->equals(hole) ) {
+                    f2->parent = f.get();
+                    break;
+                }
+            }
+        }
+    }
+}
+
+static std::unique_ptr<geom::MultiPolygon> collectFacesWithEvenAncestors(
+    std::vector<std::unique_ptr<Face>>& faces) {
+    std::vector<geom::Geometry*>* geoms = new std::vector<geom::Geometry*>();
```

Comment:
Why not use a `std::unique_ptr<std::vector<geom::Geometry*>>` here (and other similar locations). I'm guessing it's because the geometries would still be leaked on error, and `GeometryFactory` doesn't have a `GeometryFactory::createMultiPolygon(std::vector<std::unique_ptr<Geometry>>)` ?

Reply:
Yes exactly. That would cause memory leaks. I considered adding createMultiPolygon(std::vector<std::unique_ptr<Geometry>>) but didn't want to make this pull request touch too many things